nixos/zsa: init at unstable-2020-12-16

davidak opened the pull request:
###### Motivation for this change

be able to configure ZSA keyboards

Related: https://github.com/zsa/wally/issues/19

cc @spacekookie might want to test?

###### Things done

- [x] tested flashing keyboard
- [x] tested live training

--

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

davidak committed:
nixos/zsa: init at unstable-2020-12-16

add support for ZSA keyboards

Co-authored-by: Julien Debon <julien.debon@pm.me>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/250

Sir4ur0n commented:
Hi @davidak 

Thank you for this PR!
As a new owner of the ZSA Moonlander I am looking for a way to flash using Wally and use the Chrome connector. It doesn't work by default and I somehow found this PR.

1) How am I supposed to test this PR please? I tried:
* clone Nixpkgs repo (and checkout the same commit my NixOS is currently on)
* add your repo and cherry pick your commit 8a894fe9d0015f24841644a4fa97498ff9f8dd82
* add `hardware.keyboard.zsa.enable = true;` to my `configuration.nix`
* run `export NIX_PATH=nixpkgs=/path/to/my/clone/nixpkgs`
* run `sudo nixos-rebuild switch`
But I get this error:
```
error: The option `hardware.keyboard' does not exist. Definition values:
- In `/etc/nixos/configuration.nix':
    {
      zsa = {
        enable = true;
      };
    }
(use '--show-trace' to show detailed location information)
building Nix...
error: The option `hardware.keyboard' does not exist. Definition values:
- In `/etc/nixos/configuration.nix':
    {
      zsa = {
        enable = true;
      };
    }
(use '--show-trace' to show detailed location information)
building the system configuration...
error: The option `hardware.keyboard' does not exist. Definition values:
- In `/etc/nixos/configuration.nix':
    {
      zsa = {
        enable = true;
      };
    }
(use '--show-trace' to show detailed location information)
```

I have checked and the cherry-pick worked well (all files and modifications are there)m I guess I have done something else wrong but can't figure out what.

2) Would you mind updating the zsa/moonlander Git commit+hash please (and probably also the option description for improved discoverability for future users)? The one currently targeted doesn't support Moonlander udev rules.

Thank you!

Sir4ur0n commented:
Update: I have managed to test, the problem is that `export NIX_PATH=nixpkgs=/path/to/my/clone/nixpkgs` is not well taken into account during the `nixos-rebuild` step.
Anyway, by switching to `su` and repeating those steps, I managed to test this.

I updated the Git commit and SHA as such (to support Moonlander):
```diff
-    rev = "39ada98c490156afc8dd14f7efe2bf9cbf200361";
-    sha256 = "1nxral952sfdsjrfj8p22873zs89lwswlip4i9xv4a32r368h627";
+    rev = "e5dde3c700beab39fb941c6941e55535bf9b2af6";
+    sha256 = "0pkybi32r1hrmpa1mc8qlzhv7xy5n5rr5ah25lbr0cipp1bda417";
```

Note that while flashing works, I still get some errors (warnings?) during flashing (but they don't seem to disturb the flashing):
```
wally-cli moonlander_moonlander-default-layout_xbJpL_wy5xo.bin
⠏ Press the reset button of your keyboard.2020/12/17 11:06:39 handle_events: error: libusb: interrupted [code -10]
⠏ Press the reset button of your keyboard.2020/12/17 11:06:40 handle_events: error: libusb: interrupted [code -10]
 0 / 58784 [---------------------------------------------------------------------------------------------------------]   0.00%2020/12/17 11:06:40 handle_events: error: libusb: interrupted [code -10]
 58784 / 58784 [=================================================================================================] 100.00% 14s
Your keyboard was successfully flashed and rebooted. Enjoy the new firmware!
```

That being said, the USB Chrome connection now works on https://configure.ergodox-ez.com/train :heart_eyes: So thank you very much @davidak :bow: 

Once it's updated to the latest udev rules version from Wally, would someone mind reviewing/merging please? :bow:

Sir4ur0n commented on pkgs/os-specific/linux/zsa-udev-rules/default.nix:
```diff
@@ -0,0 +1,33 @@
+  src = fetchFromGitHub {
+    owner = "zsa";
+    repo = "wally";
+    rev = "39ada98c490156afc8dd14f7efe2bf9cbf200361";
+    sha256 = "1nxral952sfdsjrfj8p22873zs89lwswlip4i9xv4a32r368h627";
```
```suggestion
    rev = "e5dde3c700beab39fb941c6941e55535bf9b2af6";
    sha256 = "0pkybi32r1hrmpa1mc8qlzhv7xy5n5rr5ah25lbr0cipp1bda417";
```

Sir4ur0n commented on pkgs/os-specific/linux/zsa-udev-rules/default.nix:
```diff
@@ -0,0 +1,33 @@
+
+stdenv.mkDerivation {
+  pname = "zsa-udev-rules";
+  version = "unstable-2020-06-20";
```
```suggestion
  version = "unstable-2020-12-16";
```

Sir4ur0n commented on nixos/modules/hardware/keyboard/zsa.nix:
```diff
@@ -0,0 +1,27 @@
+      type = types.bool;
+      default = false;
+      description = ''
+        Enables udev rules for keyboards from ZSA like the ErgoDox EZ and Planck EZ.
+        You need it when you want to flash a new configuration on the keyboard
+        or use their live training in the browser.
+        Access to the keyboard is granted to users in the "plugdev" group.
+        You may want to install the wally-cli package.
```
```suggestion
        Enables udev rules for keyboards from ZSA like the ErgoDox EZ, Planck EZ and Moonlander Mark I.
        You need it when you want to flash a new configuration on the keyboard
        or use their live training in the browser.
        Access to the keyboard is granted to users in the "plugdev" group.
        You may want to install the wally-cli package.
```

davidak commented:
>the problem is that `export NIX_PATH=nixpkgs=/path/to/my/clone/nixpkgs` is not well taken into account during the `nixos-rebuild` step

you have to use `nixos-rebuild switch -I nixpkgs=/path/to/my/clone/nixpkgs`

thanks for testing and updating! i will check it out later. then you can review it again and we might find someone to merge. i think the issue was that no one had hardware to review...

Sir4ur0n commented:
@davidak Hey, I know this is open source (so motivation-, free-time- and best-effort-based) but would you mind updating this PR so it can be reviewed/merged and made available on the unstable branch? :sweat_smile: 

Let me know if there's anything I can do to help!

davidak commented:
I have accepted your changes, squashed the commits, rebased on maser and tested it.

Please test again and submit a review here, so we can get this merged.

sir4ur0n reviewed (approved):
Looks good, thank you!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/304
